### PR TITLE
Generalize decision engine to accept local assertions

### PR DESCRIPTION
**Pull request by ajreynol**
This refactors the decision engine to accept "local assertions", which are formulas whose scope is SAT-context dependent.

The functionality for such lemmas was already available via "skolem definitions" which dynamically become active. This PR makes it so that this functionality was also be used for theory lemmas.

This functionality is not yet leveraged, thus there are no behavior changes in this PR.

It also makes some minor simplifications where we assumed that "removable" lemmas were to be ignored by the decision heuristic.

### CODE REVIEW

File: src/theory/lemma_property.cpp [hunk 1]
```diff
@@ -52,11 +52,14 @@ bool isLemmaPropertyNeedsJustify(LemmaProperty p)
...
   return (p & LemmaProperty::NEEDS_JUSTIFY) != LemmaProperty::NONE;
 }
-
 bool isLemmaPropertyInprocess(LemmaProperty p)
 {
   return (p & LemmaProperty::INPROCESS) != LemmaProperty::NONE;
 }
+bool isLemmaPropertyLocal(LemmaProperty p)
+{
+  return (p & LemmaProperty::LOCAL) != LemmaProperty::NONE;
```
Comment (mpreiner) [inline src/theory/lemma_property.cpp:61]: Shouldn't this explicitly check for `(p & LemmaProperty::LOCAL) == LemmaProperty::LOCAL`? If `NONE` wouldn't be zero the above wouldn't work.

File: src/prop/theory_proxy.cpp [hunk 2]
```diff
@@ -162,15 +171,32 @@ void TheoryProxy::notifySkolemDefinition(Node a, TNode skolem)
...
   {
     return;
   }
   // notify the decision engine
-  d_decisionEngine->addAssertion(a, skolem, isLemma);
+  std::vector<TNode> assertions{a};
+  if (local)
+  {
+    // If it is marked local, add as local assertions.
+    d_decisionEngine->addLocalAssertions(assertions);
```
Comment (mpreiner) [inline src/prop/theory_proxy.cpp:189]:
```suggestion
    d_decisionEngine->addLocalAssertions({a});
```

File: src/prop/theory_proxy.cpp [hunk 3]
```diff
@@ -162,15 +171,32 @@ void TheoryProxy::notifySkolemDefinition(Node a, TNode skolem)
...
+                                  bool local)
 {
   // ignore constants
   if (a.isConst())
   {
     return;
   }
   // notify the decision engine
-  d_decisionEngine->addAssertion(a, skolem, isLemma);
+  std::vector<TNode> assertions{a};
```
Comment (mpreiner) [inline src/prop/theory_proxy.cpp:185]:
```suggestion
```

File: src/prop/theory_proxy.cpp [hunk 4]
```diff
@@ -162,15 +171,32 @@ void TheoryProxy::notifySkolemDefinition(Node a, TNode skolem)
...
+  if (local)
+  {
+    // If it is marked local, add as local assertions.
+    d_decisionEngine->addLocalAssertions(assertions);
+  }
+  else if (skolem.isNull() || !d_dmTrackActiveSkDefs)
+  {
+    // Otherwise, if it is not a skolem definition, or we are treating
+    // skolem definitions as ordinary assertions, we add it now.
+    d_decisionEngine->addAssertions(assertions);
```
Comment (mpreiner) [inline src/prop/theory_proxy.cpp:195]:
```suggestion
    d_decisionEngine->addAssertions({a});
```

File: src/prop/theory_proxy.cpp (hunk 2 above)
Comment (mpreiner) [reply src/prop/theory_proxy.cpp:189]: Avoids unnecessary creation of assertions if none of the branches are triggered.

File: src/theory/lemma_property.cpp (hunk 1 above)
Comment (ajreynol) [reply src/theory/lemma_property.cpp:61]: `NONE` has to be zero; this check is doing the same as the other properties, are you suggesting we change all of them?

I don't have a strong opinion on this.

Comment (mpreiner) [reply src/theory/lemma_property.cpp:61]: Yes, I think it'd be cleaner to change the check and not rely on NONE being zero.

Comment (ajreynol) [reply src/theory/lemma_property.cpp]: Done

**Review by mpreiner — APPROVED** [review]

File: src/decision/decision_engine.h [hunk 5]
```diff
@@ -47,23 +47,16 @@ class DecisionEngine : protected EnvObj
   /** Is the DecisionEngine in a state where it has solved everything? */
   virtual bool isDone() = 0;
   /**
-   * If skolem is null, notify this class that assertion is an (input)
-   * assertion, not corresponding to a skolem definition.
-   *
-   * If skolem is non-null, notify this class that lem is the skolem definition
-   * for skolem, which is a part of the current assertions.
+   * Adds assertions lems to satisfy that persist in the user context.
+   * All input assertions and relevant lemmas are added via this call.
```
Comment (aniemetz) [inline src/decision/decision_engine.h:51]: Documentation format (@param, @return)

File: src/decision/justification_strategy.h [hunk 6]
```diff
@@ -144,32 +144,24 @@ class JustificationStrategy : public DecisionEngine
...
+   * skolems we have yet to see in the current SAT context, where lems are the
+   * skolem definitions for each such skolem.
    */
-  bool needsActiveSkolemDefs() const override;
+  void addLocalAssertions(const std::vector<TNode>& lems) override;
 
  private:
   /**
    * Helper method to insert assertions in `toProcess` to `d_assertions` or
    * `d_skolemAssertions` based on `useSkolemList`.
```
Comment (aniemetz) [inline src/decision/justification_strategy.h:162]: Can you improve this documentation to indicate which case applies to which? Where is it added when `useSkolemList` is true?

File: src/prop/theory_proxy.h [hunk 7]
```diff
@@ -213,10 +214,15 @@ class TheoryProxy : protected EnvObj, public Registrar
   std::unique_ptr<decision::DecisionEngine> d_decisionEngine;
 
   /**
-   * Whether the decision engine needs notification of active skolem
-   * definitions, see DecisionEngine::needsActiveSkolemDefs.
+   * Whether we need to track active skolem definitions for the preregistrar
+   * or for the decision engine.
```
Comment (aniemetz) [inline src/prop/theory_proxy.h:218]:
```suggestion
   * True if we need to track active skolem definitions for the preregistrar,
   * false for the decision engine.
```

File: src/decision/justification_strategy.h (hunk 6 above)
Comment (ajreynol) [reply src/decision/justification_strategy.h:162]: It should be renamed "local".

Comment (aniemetz) [reply src/decision/justification_strategy.h]: Sure, but the main question was to clarify in which case it is added to `d_assertions`, when `local` is true?

**Pull request merged**